[2.0.x] Light-weight status screen for ST7920 based printers

marcio-ao opened the pull request:
Added an alternative light-weight status screen for the ST7920 as originally discussed in #9190

- The status screen uses the ST7920 character generator to greatly
  reduce SPI traffic and MCU load when updating the status screen.

- The status screen by-passes u8g and writes commands directly
   to the controller. All other menus will continue to use the u8g library.

- Has been tested with the RepRapDiscount Full Graphics Smart Controller
  but should work with any 128x64 LCD that uses an ST7920 or fully compatible
  controller.

![st7920](https://user-images.githubusercontent.com/29129419/36230227-f6640aae-1196-11e8-94aa-18ea8a8db0a6.png)

marcio-ao committed:
Added a light-weight status screen for the ST7920

- The status screen uses the ST7920 character generator to greatly
  reduce SPI traffic and MCU load when updating the status screen.

- Has been tested with the RepRapDiscount Full Graphics Smart Controller
  but should work with any LCD that uses an ST7920 or fully compatible
  controller.

marcio-ao committed:
Added a light-weight status screen for the ST7920

- The status screen uses the ST7920 character generator to greatly
  reduce SPI traffic and MCU load when updating the status screen.

- Has been tested with the RepRapDiscount Full Graphics Smart Controller
  but should work with any LCD that uses an ST7920 or fully compatible
  controller.

marcio-ao committed:
Lightweight status screen for ST7920

- This status screen uses the ST7920 character generator to greatly
  reduce SPI traffic and MCU load when updating the status screen.

- Has been tested with the RepRapDiscount Full Graphics Smart Controller
  but should work with any LCD that uses an ST7920 or fully compatible
  controller.

marcio-ao committed:
Added files missing from earlier commit.

marcio-ao committed:
Fixed compilation when using avr-gcc

- avr-gcc is more strict than the Arduino IDE, it seems.

thinkyhead commented:
Is this compatible with any other ST7920 displays aside from the RRD FGSC?
Perhaps even _all_ ST7920-based displays??

- `REPRAPWORLD_GRAPHICAL_LCD`
- `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER`
- `LCD_FOR_MELZI`
- `SILVER_GATE_GLCD_CONTROLLER`
- `CR10_STOCKDISPLAY`

thinkyhead committed:
Proposed changes to ST7920 lightweight status screen

marcio-ao commented:
@thinkyhead : It should be compatible with all displays that have a ST7920 or a 100% compatible chip and provided the memory map is identical (which I assume will be for all 128x64 displays). I wasn't exactly sure which these were so I only listed the RRD FGSC, but if you know these other displays use the ST7920 then in all likelihood they would work too.

Some of the displays out there, such as the MKS 12864 OLED, are 128x64 but they seem to use a chip that only has a graphics buffer. These would not be compatible and would need to continue to use the U8G-based code.

marcio-ao commented:
Incidentally, one could possibly write a software character generator that emulates the ST7920 for displays that don't have one. This would negate some of the savings in SPI traffic, but it would probably run much faster than the U8G code since the characters glyphs would be byte-aligned. If there is a strong interest in this, I could look into writing such a compatibility layer to allow the light-weight status screen to work on all 128x64 displays, even the bitmap based ones. Maybe we see how popular this new interface becomes first, as it would be a fairly complex undertaking.

ejtagle commented:
Just a point to note: The ST7920 is the slowest controller. All the others are, at least 2x faster, and OLED based ones are probably 20x faster. I am referring to the acievable SPI transfer speed. So, i guess this PR is mostly expremely valuable for the ST7920...

marcio-ao commented:
@ejtagle : That's perfect then. I started this work mostly because I wanted to learn more about the ST7920, not because I was trying to solve any particular problem; but it's nice to know I solved a problem by accident :)

thinkyhead commented:
I've enable this feature for the displays I listed above. If one of them happens not to work we're sure to find out eventually.

Please test this PR with the changes I've made. If everything functions properly then this can be merged anytime.

thinkyhead commented:
> I started this work mostly because I wanted to learn more about the ST7920

And now you've taught us all a thing or two.

One of the things I like about Marlin as a project is that it collects together so much knowledge about the boards and peripherals we all use. It's one of the reasons I prefer to preserve support for older boards and obscure machines.

That said, we still might retire the Makibox configuration for 2.1.

thinkyhead commented:
- Shorter option names `LIGHTWEIGHT_UI`, `STATUS_EXPIRE_SECONDS`.
- Moved options to `Configuration_adv.h`.
- Shorter filenames that start with `status_screen`.
- The standard status screen gets its own file too.
- Added a Travis test for the new option.
- Tweaks to spacing, code style.

thinkyhead commented:
Thinking about smarter status messages…. Some (e.g., "Heating Done") may be fine to expire after a period of time (as they can do with an `LCD_PROGRESS_BAR` sub-option). Important messages should stick around until they're seen. One thing we don't have is messages that stick around only until you click the button. That would give the best of both. Since a message could accidentally get dismissed right after it appeared, a click should not clear it until it's been up for 30 seconds.

marcio-ao commented:
@thinkyhead: Yes, I realizes some message may be important, so this is why I made the countdown only happen when there is axis motion. Generally if there is an error message that aborted the print, the motion will stop so the message will still be on the screen when the user gets back. I realize this isn't 100% foolproof, but it should cover most common cases.

marcio-ao committed:
Merge branch 'pr-lite-st7920-status' of github.com:marcio-ao/Marlin into pr-lite-st7920-status

marcio-ao committed:
Removed unused configuration options.

marcio-ao commented:
@thinkyhead : It works. Although it looks like you had moved the configs to "Configuration_adv.h" but you forgot to remove the configs I had added to the "Configuration.h" files of the sample printers. I went ahead and cleaned that up.

ejtagle commented:
@marcio-ao : I guess you have already thought about this: The ST7920 also supports 4 user-defined 16x16 characters. Those could be used to display the Heater and Fan icons, and by redefining them "on the fly", you can even create animated icons. It is interesting, as using the internal character generator means a lot less SPI traffic, as you said

AnHardt commented:
@ejtagle  
There is a efficient way for animations - but it's the other way around.
Predefine the phases of the motion in different custom characters (2x16 byte) and switch the characters (1 byte) for motion.

marcio-ao commented:
@ejtagle, @AnHardt : You both seem to have missed that in fact I am already using a single character animation frames for the fan and the heated bed animations. Here is the code that alternates the frame in DDRAM:

https://github.com/marcio-ao/Marlin/blob/aada1e698fe35abdf341f6a1f5abc622a70d44a9/Marlin/src/lcd/dogm/ultralcd_impl_st7920_lite_status_screen_impl.h#L348

The following elements are drawn to the GDRAM (bitmap RAM), generally once:

- Nozzle icon(s) (drawn once only)
- Feedrate icon (drawn once only)
- Progress bar outline and shading (updated only when progress changes)
- Bed surface without heat (drawn once only)

The following icons are pre-loaded into CGRAM (character generator RAM) at startup and then alternated using single writes to DDRAM:

1. Fan icon
2. Fan icon rotated 45 degrees
3. Heat waves facing left
4. Heat waves facing right

All the text and alternating characters for the animation are drawn to DDRAM (text buffer).

marcio-ao committed:
Fixed dual-extruder in Lightweight UI.

- Moved bed icon down when the machine is configured with two nozzles.

marcio-ao commented:
@thinkyhead : Found a bug where on the dual-nozzle setup the bed icon was getting drawn in the wrong place. Made an commit to fix that.

cjsoong commented:
Temperature can only display 0 to 255.
Can not display negative temperature.

marcio-ao commented:
@cjsoong: That's a good point. Just because I've never printed above 255, probably does not mean nobody will :) Why would we need negative temperatures, however?

cjsoong commented:
@marcio-ao 
Some temperature sensors open circuit will show negative temperature,
But "Light-weight UI" will be displayed as very high temperatures, may cause false positives user problems.
I have modified to show negative temperature
![light-weight_ui_fix](https://user-images.githubusercontent.com/26795599/36491218-2d3b27a0-1765-11e8-8aa9-4eb211f04758.jpg)

marcio-ao committed:
Changed temp type from uint8_t to int16_t

- Also deleted old implementation files which had been renamed.

marcio-ao commented:
@cjsoong : I pushed a fix. Thank you for spotting that.

Did you also change the bed icon? I notice it has corners in your screenshot :)

marcio-ao committed:
Removed unnecessary extended mode switch.

- This will slightly reduce the SPI requirements for
  updating the progress bar.

marcio-ao committed:
Added bed corners to bed icon.

- This makes it look less like a dash (idea stolen from cjsoong).

thinkyhead committed:
Add LIGHTWEIGHT_UI to example configs

thinkyhead commented:
Rebasing and squashing… Be sure to get a fresh copy before adding on.

cjsoong commented:
@marcio-ao 
bed icon A little change, this will not look like a minus sign

thinkyhead commented:
I've been playing with graphics lately. Some bed icon concepts:
![image](https://user-images.githubusercontent.com/698003/36526918-8de23462-1775-11e8-8b39-834895416dc0.png)

Four- and three-frame fan animations:
![image](https://user-images.githubusercontent.com/698003/36527019-f46edd7a-1775-11e8-8d75-a0f18cc9315e.png)

![image](https://user-images.githubusercontent.com/698003/36527026-ff05d158-1775-11e8-82db-08f0281946cc.png)

marcio-ao commented:
@cjsoong : I made a commit with your modified bed icon. It does help make it look less like a plus sign. Thank you! One thing that's nice is that in the "lightweight" UI, the bed icon is underneath the nozzle, so it's a bit more visually unambiguous than it is in the standard UI screen, where they are side-by-side.

marcio-ao commented:
@thinkyhead : I like the fan icons with more frames, but unfortunately for maximum efficiency on the ST7920, I only have four custom characters to work with, and right now they are split between the fan and the heated bed, so just two frames a piece. A perspective bed icon might look decent on the lightweight UI, but there are fewer pixels to work with, so I'm not sure how good it'll look until I try scaling it down. I'll try it out.

marcio-ao committed:
Merge remote-tracking branch 'origin/pr-lite-st7920-status' into pr-lite-st7920-status

marcio-ao commented:
Nope. Doesn't look good at all, not enough pixels. I think @cjsoong has the best icon so far for the lightweight GUI.

![parallelogram1](https://user-images.githubusercontent.com/29129419/36546480-c623c678-17a8-11e8-9e0e-fe05f8757ebb.JPG)
![parallelogram2](https://user-images.githubusercontent.com/29129419/36546486-c7b4ae8a-17a8-11e8-9f29-dbc6ea0a007f.JPG)

thinkyhead commented:
Actually, with the nozzle over the parallelogram, you can totally grok in-context that it's the bed.

thinkyhead commented:
> I only have four custom characters to work with, and right now they are split between the fan and the heated bed, so just two frames a piece.

What happens if you use 1 character, but you redefine its bitmap between screen renderings? If that can be done then it would allow for any number of frames. Is that allowed during operation, or does it have to be part of the initialization?

AnHardt commented:
> What happens if you use 1 character, but you redefine its bitmap between screen renderings?

Than you have to transfer 16 bytes per phase instead ot 1. The same amount as when drawing the icon in graphic-mode. 

Running in circles. 

For my taste the vertical lines in the FR-icon could be a bit fatter to match the on chip font. It looks a bit 'skinny'.

thinkyhead commented:
> Then you have to transfer 16 bytes per phase…

I see. Of course, it could be done at any point in the lengthy 1s downtime between screen updates.

> For my taste the vertical lines in the FR-icon could be a bit fatter to match the on chip font. It looks a bit 'skinny'.

Not a big Jony Ive fan? Me neither. And I agree, it would look better with a little weight.

marcio-ao commented:
@thinkyhead, @AnHardt : It's fairly easy to change the icons. I think at one point I had a fatter FR icon, but I wasn't satisfied with how it looked, so I gave up on making one. Can't say I am a fan of the parallelogram for the bed, but again, simple to change if that is desired.

Thank you for merging this in! :)

thinkyhead commented:
In future we'll have more configurable display options. I like the flat bed icon well enough that I don't feel the need to go all 3D. (I'd need to design a 3D nozzle first!)

CdRsKuLL commented:
Hi, I was wondering if you had any idea why my screen is doing the following... 

https://www.youtube.com/watch?v=yh-PQ7BpZl8

It was fine in 1.1.8 but now I've moved everything over to 1.1.9 the screen isn't right. However, if I enable Light-weight option, it displays the main info screen correctly but not the menus.

Any ideas?  

Thanks,

thinkyhead commented:
@CdRsKuLL — See https://github.com/MarlinFirmware/Marlin/issues/3815#issuecomment-412619421

marcio-ao commented:
The Klipper FW is using a port of the lighweight UI. There are some cool new icon concepts being discussed there:

https://github.com/KevinOConnor/klipper/issues/652#issue-360570315

thinkyhead commented:
Interesting! Checking it out…

fiveangle commented on Marlin/Configuration_adv.h:
```diff
@@ -656,6 +656,22 @@
   // Swap the CW/CCW indicators in the graphics overlay
   //#define OVERLAY_GFX_REVERSE
 
+  #if ENABLED(U8GLIB_ST7920)
```
Would it make more sense to put the check for a U8GLIB_ST7920-based display into the LIGHTWEIGHT_UI code, then put a "LIGHTWEIGHT_UI requires a U8GLIB_ST7920 based graphical display." into SanityCheck.h ?

How will users know they have a U8GLIB_ST7920-based display without digging deeper beyond both Configuration*.h and the board pin*.h files ?

(I mention this because I ran into this confusion by setting LIGHTWEIGHT_UI on what I thought was a U8GLIB_ST7920 based LCD and not seeing any change, but turns out it was an U8GLIB_SSD1306 LCD).

thinkyhead replied:
The configurator uses these `#if` wrappers to decide whether to present the option. But we should also add a sanity check.
